### DISCUSSION TITLE
How do I pass compiler/linker arguments to Emscripten?

### DISCUSSION THREAD

**Discussion by ahmedyarub** [microsoft/vcpkg#30108] [Q&A]
I'm creating a cross-platform application that uses Qt downloaded using the official tool in addition to some other libraries that I could successfully build in vcpkg like OpenSSL, libssh, and aws-sdk-cpp. Now I can build the application using the single-threaded Qt 6.5.0 beta and link them to vcpkg libraries. However, I need the Qt Concurrent library which means that I have to build against the multi-threaded version of Qt. Linking to these Qt libraries does work but when I add my vcpkg libraries I get the following error:

> wasm-ld: error: --shared-memory is disallowed by log.c.o because it was not compiled with 'atomics' or 'bulk-memory' features.

That error should be resolved by passing -pthread parameter, -matomics, or some other compiler/linker parameters to Emscripten. I did try to pass them like this in a new toolchain file but that still didn't work:

> set(VCPKG_CXX_FLAGS "-matomics -s SHARED_MEMORY=1")
> set(VCPKG_C_FLAGS "-matomics -s SHARED_MEMORY=1")
> set(VCPKG_LINKER_FLAGS "-matomics -s SHARED_MEMORY=1")

Any ideas?

**Comment by autoantwort** [comment]
`VCPKG_CXX_FLAGS` et al are triplets variables, you have to set them in a new triplet. https://learn.microsoft.com/en-us/vcpkg/users/triplets#vcpkg_cxx_flags

**Reply by ahmedyarub** [reply]
I did create a new triplet and added the parameters like this to the end of a new triplet but that didn't change anything:
```
set(VCPKG_CXX_FLAGS "-pthread")
set(VCPKG_C_FLAGS "-pthread")
```
I then tried to manually build libssh outside vcpkg passing the parameters as command-line parameters and it worked as expected:
`-DCMAKE_CXX_FLAGS="SHELL:-pthread" -DCMAKE_C_FLAGS="-pthread"`

Is there a bug in vcpkg or am I doing something wrong?

**Reply by ahmedyarub** [reply]
I noticed that flag was nowhere to be found in the generated build scripts so I tried adding the lines to the toolchain of Emscripten instead. I could see the flag added to the scripts but the files were still not built for pthreads.

**Comment by netcorefan1** [comment]
@ahmedyarub I'm also having the same problem. The docs says that `set(VCPKG_*_FLAGS .)` only works when not using `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` and according to this [article](https://decovar.dev/blog/2022/10/30/cpp-dependencies-with-vcpkg/#enabling-pthreads-support) written by @retifrav  (#3577), it is expected to work when using any of `/path/to/vcpkg/scripts/toolchains/*.cmake`. In my case I'm just using the default one (`vcpkg.cmake`), but the flags are just ignored.

_Did you found a workaround?_

> I then tried to manually build libssh outside vcpkg passing the parameters as command-line parameters and it worked as expected: `-DCMAKE_CXX_FLAGS="SHELL:-pthread" -DCMAKE_C_FLAGS="-pthread"`
> 
> Is there a bug in vcpkg or am I doing something wrong?

Same for me and I'm also wondering if this is a bug that should be reported.

**Reply by ahmedyarub** [reply]
> @ahmedyarub I'm also having the same problem. The docs says that `set(VCPKG_*_FLAGS .)` only works when not using `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` and according to this [article](https://decovar.dev/blog/2022/10/30/cpp-dependencies-with-vcpkg/#enabling-pthreads-support) written by @retifrav (#3577), it is expected to work when using any of `/path/to/vcpkg/scripts/toolchains/*.cmake`. In my case I'm just using the default one (`vcpkg.cmake`), but the flags are just ignored.
> 
> _Did you found a workaround?_
> 
> > I then tried to manually build libssh outside vcpkg passing the parameters as command-line parameters and it worked as expected: `-DCMAKE_CXX_FLAGS="SHELL:-pthread" -DCMAKE_C_FLAGS="-pthread"`
> > Is there a bug in vcpkg or am I doing something wrong?
> 
> Same for me and I'm also wondering if this is a bug that should be reported.

Unfortunately, the only workaround I could find is modifying the toolchain of Emscripten that resides inside the emsdk folder.
Far from being ideal but it worked for me.

**Reply by netcorefan1** [reply]
Did you tried to directly deal with `VCPKG_CMAKE_CONFIGURE_OPTIONS` ? I had [success](https://github.com/microsoft/vcpkg/discussions/26050#discussioncomment-5585021) to make CMake recognize my custom flags (although I got a different output than expected).

**Comment by fwcd** [comment]
I have had success defining a custom triplet with

```cmake
set(ENV{EMCC_CFLAGS} "$ENV{EMCC_CFLAGS} -pthread")
```

This is certainly not ideal, but seems to work (you can verify that the generated objects indeed have this enabled by checking `wasm-objdump -x <path to object file>`, which should include `[+] atomics`).
